### DISCUSSION TITLE
Split `khepri_fun` into its own library?

### DISCUSSION THREAD

**Discussion by the-mikedavis** [rabbitmq/khepri#60] [Ideas]
I was parousing through the Erlang Forums post and I caught @dumbbell's [comment](https://erlangforums.com/t/khepri-a-tree-like-replicated-on-disk-database-library-for-erlang-and-elixir-introduction-feedbacks/438/29?u=the-mikedavis) (btw congrats on 0.2 :) that it might be nice to have khepri_fun as its own library.

I have been considering opening up this discussion as well because I have a use-case for khepri_fun unrelated to the database aspects of khepri.

<details><summary>My use case...</summary>

is a bit strange admittedly 😄. We have business logic code that can be expressed in a purely-functional transaction-like way and then other code that provides an impure runtime for that code (Phoenix endpoints, database connections, HTTP clients, etc.). I would like to lint the logic code for impure instructions and also re-compile it into a fully self-contained in a module that I can hand over to the runtime code (with `code:load_binary/3` for example). The modules being both fully self-contained and purely functional means that we could load multiple versions of the same logic code into memory at once and compare outputs given the same inputs (only performing side-effects on the outputs for the "live" version), so you can expose new changes in business logic to production traffic safely and quickly. A big plus is that this approach works for any beam language 🎉

I have a working prototype of this in Elixir (eventually I wanted to rewrite in Erlang and open-source), and its usage looks something like

```elixir
{module, beam} = Overhaul.compile(%{{Kernel, :+, 2} => :add})
{:module, ^module} = :code.load_binary(module, to_charlist(inspect(__MODULE__)), beam)
module.add(2, 3)
#=> 5
```

I can talk about this more if you're interested (and I can probably give you read access to `overhaul`) but in my opinion it is pretty niche 😅

<hr>

</details>

What would you think about separating out khepri_fun into its own library focused on linting and re-organizing beam modules and functions?

**Comment by dumbbell** [comment]
Hi!

That's a good use case as well, I didn't think about it before. Nothing strange there :-)

And yes, `khepri_fun` needs to make it into its own library. It could benefit others outside of Khepri, which would be awesome! It also became too big: I need to split it into multiple modules.

There is a significant amount of work with the Khepri testsuite too. It started as a testsuite to make sure transaction functions didn't do anything forbidden. It's now a large testsuite testing many things, sometimes unrelated to transactions.

But the most difficult part remains the same as with every new projects: finding a name! A name generator just suggested me *Qyrad The Taker Of Life* or *Resosdienth The Insane*. I like them and the domain names are available! :-) More seriously, I will come up with some ideas. @kjnilsson: You are good at picking names, what would you suggest for the function extraction code of Khepri? @the-mikedavis, feel free to share yours as well!

**Reply by the-mikedavis** [reply]
Surely _Qyrad The Taker Of Life_ must be taken on Hex already 😄

I thought about this a bit but didn't find anything I really liked. There were some very "beam"-related ones like `ray` or `laser` that I was surprised haven't been taken yet on Hex. Maybe something along those lines like `refractor`/`mirror`? Or maybe something more math-y like `isomorph`? (although I don't know for sure I'm using _isomorphic_ correctly there)

I'm not in love with any of these yet - I'm looking forward to what y'all come up with

**Comment by dumbbell** [comment]
Quick updated: the code was shipped as "Horus" (https://github.com/rabbitmq/horus) a long time ago.
